### PR TITLE
drivers: video: Conversion of k_work API

### LINKED ISSUE #34090: Convert video_sw_generator.c to new kwork API
Convert drivers/video/video_sw_generator.c to use new kwork API.

See the following issue for details:

https://github.com/zephyrproject-rtos/zephyr/issues/33104

### PR DESCRIPTION
**Pull request by loicpoulain**
Replace deprecated API with the recommended alternative.

Fixes: #34090
Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

### CODE REVIEW

**Review by pabigot — APPROVED** [review]
This is OK from the perspective of changing the API.

The race conditions in the driver are unchanged: issuing an unchecked work cancel does not eliminate the data race between the work handler and other threads accessing the generator data.

File: drivers/video/video_sw_generator.c [hunk 1]
```diff
@@ -56,14 +56,14 @@ static int video_sw_generator_stream_start(const struct device *dev)
...
-	return k_delayed_work_submit(&data->buf_work, K_MSEC(33));
+	return k_work_schedule(&data->buf_work, K_MSEC(33));
 }
 
 static int video_sw_generator_stream_stop(const struct device *dev)
 {
 	struct video_sw_generator_data *data = dev->data;
 
-	k_delayed_work_cancel(&data->buf_work);
+	k_work_cancel_delayable_sync(&data->buf_work);
```
Comment (pabigot) [inline drivers/video/video_sw_generator.c:66]: Changing to `_sync` doesn't solve the problem: a cancel can fail, and if it does you have a race condition.

**Comment by loicpoulain** [comment]
> This is OK from the perspective of changing the API.
> 
> The race conditions in the driver are unchanged: issuing an unchecked work cancel does not eliminate the data race between the work handler and other threads accessing the generator data.

I've replaced cancel with cancel_sync to ensure it is not running once stopped (until it is restarted). Not sure about what race you're pointing here, the work mostly play with fifo_get/put which is AFAIK thread safe. Other elements like video data pattern are not 'changed' while the video stream (and so the work) is running (started).

File: drivers/video/video_sw_generator.c (hunk 1 above)
Comment (pabigot) [reply drivers/video/video_sw_generator.c:66]: Re-reading my own documentation: that's not true; the work item will be idle when this completes.  Note that `stop()` can no longer be invoked from an ISR, though.

Comment (loicpoulain) [reply drivers/video/video_sw_generator.c:66]: > Re-reading my own documentation: that's not true; the work item will be idle when this completes.

Ok, good, that was my understanding ;-)

> Note that `stop()` can no longer be invoked from an ISR, though.

Yes, that's fine, both video stream start() and stop() should be (and are in current usages) executed in process context.

Regards,
Loic

**Review by galak — APPROVED** [review]

**Pull request merged**